### PR TITLE
Make drain() consistently asynchronous.

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

File: testing/dart/channel_buffers_test.dart [hunk 1]
```diff
@@ -30,6 +32,32 @@ void main() {
...
+    await completer.future;
+    log.add('after await');
+    expect(log, <String>[
+      'before drain',
+      'callback',
+      'after drain, before await',
+      'before drain, microtask',
+      'after await',
+    ]);
+  });
```
Comment (Hixie) [inline testing/dart/channel_buffers_test.dart:59]: actual output is:

              'before drain',
              'after drain, before await',
              'before drain, microtask',
              'callback',
              'after await'

Comment (Hixie) [reply testing/dart/channel_buffers_test.dart:59]: ok i just forgot to recompile my code, never mind...

**Comment by Hixie** [comment]
Test failures will be addressed by https://github.com/flutter/flutter/pull/65579

**Comment by Hixie** [comment]
This is blocked because it exposes a number of issues with ChannelBuffers that result in an infinite loop in one of the framework tests.

1. drain is async, but if a message comes in while draining, it might be delivered before drain is finished (unordered) -- possible solution is to push messages onto the queue if they come in when the queue is non-empty.
2. setting a handler to null still drains the queue -- this should just not happen.
3. drain will handle messages that are added while it's running, so you can end up in an infinite loop, which is especially noticeable if the handler is removed while you're draining since then the framework will just keep putting each message back onto the queue.
4. drain being called twice at the same time will result in two simultaneous drains to different callbacks.

We could avoid all of these by removing the concept of a ring buffer (https://github.com/flutter/flutter/issues/66071), but that's a potentially very breaking change.

**Comment by Hixie** [comment]
https://github.com/flutter/flutter/pull/66073 probably unblocks this enough for now.

**Comment by Hixie** [comment]
cc @gaaclarke @yjbanov This is ready for review now, with the framework-side changes having landed.

**Review by gaaclarke — APPROVED** [review]

**Comment by Hixie** [comment]
This failed in google3, will follow-up in https://github.com/flutter/flutter/issues/66699.

**Pull request merged**